R/expr-expr.R

```diff
@@ -70,195 +69,523 @@ pl__deserialize_expr <- function(data, ..., format = c("binary", "json")) {
   })
 }
 
+#' Add two expressions
+#'
+#' Method equivalent of addition operator `expr + other`.
+#' @param other numeric or string value; accepts expression input.
+#' @return [Expr][expr__class]
+#' @seealso
+#' - [Arithmetic operators][S3_arithmetic]
+#' @examples
+#' df <- pl$DataFrame(x = 1:5)
+#'
+#' df$with_columns(
+#'   `x+int` = pl$col("x")$add(2L),
+#'   `x+expr` = pl$col("x")$add(pl$col("x")$cum_prod())
+#' )
+#'
+#' df <- pl$DataFrame(
+#'   x = c("a", "d", "g"),
+#'   y = c("b", "e", "h"),
+#'   z = c("c", "f", "i")
+#' )
+#'
+#' df$with_columns(
+#'   pl$col("x")$add(pl$col("y"))$add(pl$col("z"))$alias("xyz")
+#' )
 expr__add <- function(other) {
   wrap({
     other <- as_polars_expr(other, as_lit = TRUE)
     self$`_rexpr`$add(other$`_rexpr`)
   })
 }
 
+#' Substract two expressions
+#'
+#' Method equivalent of subtraction operator `expr - other`.
+#' @inherit expr__truediv params return
+#' @seealso
+#' - [Arithmetic operators][S3_arithmetic]
+#' @examples
+#' df <- pl$DataFrame(x = 0:4)
+#'
+#' df$with_columns(
+#'   `x-2` = pl$col("x")$sub(2),
+#'   `x-expr` = pl$col("x")$sub(pl$col("x")$cum_sum())
+#' )
 expr__sub <- function(other) {
   wrap({
     other <- as_polars_expr(other, as_lit = TRUE)
     self$`_rexpr`$sub(other$`_rexpr`)
   })
 }
 
+#' Multiply two expressions
+#'
+#' Method equivalent of multiplication operator `expr * other`.
+#' @inherit expr__truediv params return
+#' @seealso
+#' - [Arithmetic operators][S3_arithmetic]
+#' @examples
+#' df <- pl$DataFrame(x = c(1, 2, 4, 8, 16))
+#'
+#' df$with_columns(
+#'   `x*2` = pl$col("x")$mul(2),
+#'   `x * xlog2` = pl$col("x")$mul(pl$col("x")$log(2))
+#' )
 expr__mul <- function(other) {
   wrap({
     other <- as_polars_expr(other, as_lit = TRUE)
     self$`_rexpr`$mul(other$`_rexpr`)
   })
 }
 
-expr__true_div <- function(other) {
+#' Divide two expressions
+#'
+#' Method equivalent of float division operator `expr / other`.
+#'
+#' Zero-division behaviour follows IEEE-754:
+#' - `0/0`: Invalid operation - mathematically undefined, returns `NaN`.
+#' - `n/0`: On finite operands gives an exact infinite result, e.g.: ±infinity.
+#' @inherit expr__add return
+#' @param other Numeric literal or expression value.
+#' @seealso
+#' - [Arithmetic operators][S3_arithmetic]
+#' - [`<Expr>$floor_div()`][Expr_floor_div]
+#' @examples
+#' df <- pl$DataFrame(
+#'   x = -2:2,
+#'   y = c(0.5, 0, 0, -4, -0.5)
+#' )
+#'
+#' df$with_columns(
+#'   `x/2` = pl$col("x")$div(2),
+#'   `x/y` = pl$col("x")$div(pl$col("y"))
+#' )
+expr__truediv <- function(other) {
```

Comment:
Why did you change the name to `truediv`?
IIUC this has a different name in Rust and Python, so we have to choose which one we prefer.
I think with underscore makes more sense here.

Reply:
`truediv` comes from Python. Rust and R do not have to choose this.
https://docs.python.org/3/library/operator.html

Reply:
I think we should just match py-polars and rename `floor_div` to `floordiv`, and use `truediv`. If we had `floor_div()` / `true_div()` in R, we probably would have matched those but given that those don't exist I think we should prioritize consistency between r-polars and py-polars functions. It would just be annoying to have an error when converting R <-> python code just because of an underscore.

> Python Polars cannot use not(), so it adopts not_(), and so on, which is a limitation of the Python world of naming.

IMO that's a different issue. For division, they could have used `floor_div()` and `true_div()` but just decided to stick to the most familiar in python (unless I missed something).

Reply:
Ok. I think it is not a bad idea to follow Python here.

Reply:
On second thought, I still think this method name should be changed to `true_div` and `floor_div` as it seems odd even for Python Polars (pola-rs/polars#19531).

Reply:
@etiennebacher After all, this seems to be Python-only, so wouldn't it make more sense to prefer a naming convention that uses underscores in R?
https://github.com/pola-rs/polars/issues/19531#issuecomment-2449458291

Reply:
I still think the same as above: given that `floordiv`/`truediv` don't exist in R, I don't see a reason for diverging from the py-polars syntax (especially since the syntax of function names in R varies a lot).

Reply:
So what I am saying is that since this is not Python, shouldn't Polars naming convention take precedence over Python? See this comment.
https://github.com/pola-rs/polars/issues/6120#issuecomment-1374911565

In fact, I don't think Rust follows Python's naming conventions.

Reply:
I get your point, but I think it's more important to have consistency with the actual functions implemented in py-polars. It would be weird to say that we have this inconsistency because we follow the polars naming convention while the original doesn't.

Reply:
In that case, why not use true_div and floor_div as canonical (i.e., we need to use this one in the document) and also add truediv and floordiv as aliases for compatibility with Python Polars?

Reply:
Done.